app/src/gui/components/record/form.tsx

```diff
@@ -987,286 +1017,263 @@ class RecordForm extends React.Component<any, RecordFormState> {
-                this.props.record_id
-              );
-              // if this is not a child record then is_direct will be false
-              if (is_direct === false) {
-                // publish and close
-                if (is_close === 'close') {
-                  this.props.navigate(
-                    ROUTES.INDIVIDUAL_NOTEBOOK_ROUTE +
-                      this.props.serverId +
-                      '/' +
-                      this.props.project_id
-                  ); //update for save and close button
-                  window.scrollTo(0, 0);
-                  return hrid;
-                  // publish and new record
-                } else if (is_close === 'new') {
-                  //not child record
-                  setSubmitting(false);
-                  this.props.navigate(
-                    ROUTES.INDIVIDUAL_NOTEBOOK_ROUTE +
-                      this.props.serverId +
-                      '/' +
-                      this.props.project_id +
-                      ROUTES.RECORD_CREATE +
-                      this.state.type_cached
+    );
+  }
+
+  /**
+   * Wrapper around the navigate method - navigate and scroll to top
+   * @param destination route to navigate to
+   * @param state state to be passed to navigate, optional
+   */
+  navigateTo(destination: string, state?: any) {
+    this.props.navigate(destination, state ? {state} : undefined);
+    window.scrollTo(0, 0);
+  }
+
+  /**
+   * A helper method to generate the navigation routes used after records are saved
+   * @param type - type of route we are generating (project, record, create)
+   * @param params - parameters needed for the route,
+   *          for 'project' this can be empty
+   *          for 'create' this should include formType (eg. the value of this.state.type_cached)
+   * @returns {string} route path to be passed to navigateTo
+   */
+  getRoute(
+    type: 'project' | 'create',
+    params: {
+      recordId?: string;
+      revisionId?: string;
+      formType?: string | null;
+    } = {}
+  ) {
+    const {serverId, project_id} = this.props;
+    const base = ROUTES.INDIVIDUAL_NOTEBOOK_ROUTE + serverId + '/' + project_id;
+
+    switch (type) {
+      case 'project':
+        return base;
+      case 'create':
+        if (!params.formType)
+          throw new Error('Form type required for create route');
+        return base + ROUTES.RECORD_CREATE + params.formType;
+      default:
+        return base;
+    }
+  }
+
+  /**
+   * Handle redirect after a record is saved
+   * - navigates to the appropriate place based on the user selection
+   *   in is_close, passing the right context as needed.
+   *
+   * @param ids - record and revision ids
+   * @param is_close - close or new
+   * @param setSubmitting - function to set the submitting state
+   * @returns revision id of the recently saved child record
+   */
+  redirectAfterSave(
+    ids: RecordIdentifiers,
+    is_close: string,
```

Comment:
if this is 'close' or 'new' please type it as so i.e. `is_close: 'close' | 'new'`